client/src/cbltest/api/couchbaseserver.py

```diff
@@ -593,3 +663,383 @@ def stop_xcdr(self, target: "CouchbaseServer", bucket_name: str) -> None:
+                        f"http://{self.__hostname}:8091/controller/rebalance",
+                        data=data,
+                    )
+                    resp.raise_for_status()
+                    break
+                except Exception:
+                    if attempt < 4:
+                        time.sleep(1)
+                    else:
+                        raise
+
+            # Wait for rebalance to complete
+            self._wait_for_rebalance_completion(session)
+
+    def rebalance(self, eject_failed_nodes: bool = False) -> None:
+        """
+        Rebalances the cluster in its current state.
+        Useful after failover to activate replica vBuckets.
+
+        :param eject_failed_nodes: If True, removes failed nodes from cluster.
+                                   If False, keeps them for later recovery.
+        """
+        with requests.Session() as session:
+            session.auth = (self.__username, self.__password)
+
+            # Get all node OTP IDs
+            resp = session.get(f"http://{self.__hostname}:8091/pools/default")
+            resp.raise_for_status()
+            pool_data = resp.json()
+
+            known_nodes = []
+            ejected_nodes = []
+
+            for node in pool_data.get("nodes", []):
+                otp_node = node.get("otpNode")
+                cluster_membership = node.get("clusterMembership")
+
+                # knownNodes must include ALL nodes (even failed ones)
+                known_nodes.append(otp_node)
+
+                # Only eject if explicitly requested
+                if eject_failed_nodes and cluster_membership == "inactiveFailed":
+                    ejected_nodes.append(otp_node)
+
+            # Start rebalance
+            rebalance_data = {"knownNodes": ",".join(known_nodes)}
+            if ejected_nodes:
+                rebalance_data["ejectedNodes"] = ",".join(ejected_nodes)
+
+            resp = session.post(
+                f"http://{self.__hostname}:8091/controller/rebalance",
+                data=rebalance_data,
+            )
+
+            if not resp.ok:
+                error_msg = f"Rebalance failed: {resp.status_code} - {resp.text}"
+                raise CblTestError(error_msg)
+
+            resp.raise_for_status()
+
+            # Wait for rebalance to complete
+            self._wait_for_rebalance_completion(session)
+
+    def failover(self, node_to_failover: "CouchbaseServer") -> None:
+        """
+        Performs a hard failover on a node in the cluster (simulates sudden node failure).
+
+        :param node_to_failover: The node to failover
+        """
+        with requests.Session() as session:
+            session.auth = (self.__username, self.__password)
+
+            # Get all node OTP IDs
+            resp = session.get(f"http://{self.__hostname}:8091/pools/default")
+            resp.raise_for_status()
+            pool_data = resp.json()
+
+            # Find the OTP node ID for the node to failover
```

Comment:
It seems like you perform this logic a lot.  Maybe this could be refactored into its own internal method (getting the node that you want out of the cluster)